### DISCUSSION TITLE
Use docker secret (or alternative solution) to store IB password

### DISCUSSION THREAD

**Discussion by benjaminSTW** [gnzsnz/ib-gateway-docker#93] [Q&A] [answered]
Hello,

Currently, I am using the Ib-gateway-docker on paper. When  When I will use live data, I would prefer storing my IB password in a docker secret (or similar solution) instead of having it in the .env file.

Has anyone already use docker secret with ib-gateway docker ?

Thank you for your help.

**Comment by gnzsnz** [comment]
This is not implemented currently, but I do believe that is a good idea to look at implementation options. PRs are always welcome.

To have this working would require (i'm thinking out loud here):
- support both secrets and env variables.
- secret should take precedence over env variable for passwords (tws api, vnc, ssh passphrase)
- if both are defined is a misconfiguration, stop container.
- `envsubst` is used to apply settings on IBC's config.ini, so the password will be stored there. which is the same case than right now. the thing is that envsubst expect the pass in a env var. need to test if something like `IbPassword=$(</run/secrets/tws_pass)` would work.
- `envsubst` still expects an env variable if pass is stored in var and not a secret (see first point). which opens a challenge, because we should store the secret into a var to pass it to envsubst. or we store it temporary in a var if a secret is used and unset the var after config.ini is defined.

It's doable, but will take some time. it should be possible to create a common function to manage all cases. As said, PRs are more than welcome.

**Comment by gnzsnz** [comment]
postgres image has a function do most of the work

https://github.com/docker-library/postgres/blob/44ef8b226a40f86cf9df3f9299067db6779a3aa3/docker-entrypoint.sh#L5

still need to figure out if envsubst will work

**Comment by gnzsnz** [comment]
@benjaminSTW  I'm doing some testing and trying to evaluate the "relative" security of using secrets. And I have to say that I was expecting more.

I have setup a postgres container, basically all official images follow the same approach. they have a `XX_PASSWORD_FILE` that points to the secret. however they copy the content of `XX_PASSWORD_FILE` into `XX_PASSWORD`. so for example for postgres it's `POSTGRES_PASSWORD_FILE`

after setting a docker compose file, set a secret, and running the container. i can see the content of the password

```bash
docker compose up -d postgres
docker inspect postgres | grep PASSWORD
      "POSTGRES_PASSWORD=foryoureyesonly"
```

so yes there is a security improvement by not setting the password in an `.env` file, but still the password is visible from the host.

In any case take into account that the password will be stored in IBC's config.ini file. That's required by IBC, so there not much room for improvement there.

**Comment by benjaminSTW** [comment]
hello,
thank you for your feedback.

Switching to crypted password is not that easy ! 

I can have look and try to run the script with a password stored in the docker secret. When looking at the documentation, it seems to be secure and sensitive data are not visible.

For IBC,  indeed, the password will still be available in the config.ini. I can raise the point to IBC community. 
I may be not the only one afraid of leaving my password in a config file.

PS : thank you for your work and your quick feedback !

**Reply by gnzsnz** [reply]
@benjaminSTW could you please check issue #93

I have created a pull request with secret support. and I would appreciate if you could test it and provide feedback.

I have used postgresql bash function to read a secret if it exists, and added an extra step to unset the variable with the password once it's used (postgresql does not perform this last step). additionally i have changed config.ini permissions, now it's 600. So I think this is as secure as it can get within the container.

in order to test you will need to pull from https://github.com/gnzsnz/ib-gateway-docker/tree/secret and create a docker-compose with secrets on it.

instruction on the updated README.md https://github.com/gnzsnz/ib-gateway-docker/blob/secret/template_README.md#credentials

**Comment by gnzsnz** [comment, marked as answer]
moved as issue, #96
